[4.x] Improve container extensibility

voronkovich opened the pull request:
The current design of the `Container` class lacks extensibility. For example, it's not possible to add a new type of service (e.g. [lazy services](https://github.com/joomla-framework/di/pull/58)) without dirty hacks or BC breaks.

### Summary of Changes

This PR replaces the use of `$mode`, `$shared` and `$protected` parameters with a single plain `$options` array.

```php
$container->set(
    'foo',
    fn () => new \stdClass(),
    [ 'shared' => true, 'protected' => true ]
);

$container->share(
    'foo',
    fn () => new \stdClass(),
    [ 'protected' => true ]
);

$container->protect(
    'foo',
    fn () => new \stdClass(),
    [ 'shared' => true ]
);
```

It allows us to easily add new types of services and options in the future:

```php
$container->set(
    'foo',
    fn () => new \stdClass(),
    [ 'shared' => true, 'protected' => true, 'lazy' => true ]
);
```

### Testing Instructions

Launch `phpunit` and ensure that the tests are passed.

### Documentation Changes Required

Yes

voronkovich committed:
Improve container extensibility

HLeithner commented:
I'm not really happy with the direction, you can't really test this in you own component if you have a typo or not (at least it's harder) also static code analyzers need much more support for this. It's also a hard b/c break, I'm pretty sure that this will not happen in this way.

voronkovich committed:
Throws an exception if unknown option provided

voronkovich committed:
Avoid BC break

voronkovich commented:
@HLeithner, thanks for your review!

I've added checking for allowed options. Now, if an unknown option is provided, an exception is thrown.

```
Unknown option "unshare" given. Allowed options: shared, protected.
```

Also, I've implemented a simple BC layer to avoid breaking existing code.
